working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+# Proposal for adding `FinalizationRegistry` and `WeakRef` to core libraries.
+
+Author: vegorov@google.com<br>Version: 0.1
+
+## Background
+
+This proposal includes two interconnected pieces of functionality:
+
+- ability to create _weak references_ (`WeakRef`);
+- ability to associate finalization actions with objects
+  (`FinalizationRegistry`).
+
+### Weak references
+
+_Weak reference_ is a object which holds to its _target_ in a manner which does
+not prevent runtime system from reclaiming the referenced object when it becomes
+_weakly reachable_ or, in other words, reachable only through weak references.
+When an object is dimmed _weakly reachable_ all `WeakRef` objects referencing it
+are cleared and the object is eligible for reclamation.
+
```

Comment:
JavaScript supports WeakRef as well as FinalizationRegistry. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef
With WeakRef and FinalizationRegistry you can build useful thinks like a WeakMap with iterable keys that are useful for client side memory leak detection and other use cases where you need to be able to iterate through the keys or values in a WeakMap.

Reply:
`WeakRef` as defined by ECMA-262 takes special precautions to avoid making GC too observable, see [Processing Model of WeakRef and FinalizationRegistry Objects](https://tc39.es/ecma262/#sec-weakref-invariants): `WeakRef.prototype.deref` (see [WeakRefDeref](https://tc39.es/ecma262/#sec-weakrefderef)) adds reference's target to a special list of objects which will be cleared on the turns of the event loop. This guarantees that synchronous code will never observe the GC.

I have not added anything like that into this proposal because I personally don't understand the need to overcomplicate things for the purpose of hiding GC away (and I don't want weak reference targets to be kept alive until the next event loop turn for no good reason). As I understand it most of the arguments around GC observability in TC-39 used to be centered around security, but I don't think we are in the same type of environment to worry about it ourselves. (/cc @syg - Shu, maybe you have the background on why `WeakRef` clearing was specified the way it was?) 

@jacob314 Jacob, how important `WeakRef`'s are for you usecase? If they are not important then we could pull them out of the proposal and focus on getting `FinalizationRegistry` over the finishing line.

Reply:
I think my worry is mainly over having an object which changes *state* due to GC. That's a change in state that is extra-language since it doesn't come from executing code.
The finalization registry callback can change a variable from `true` to `false` too, but the state change comes from running Dart code. It *feels* different.

In practice, no-one can see whether the `WeakRef` has an internal finalization callback function which sets its state to `null` when the GC recognizes the object as otherwise unreachable, and if that code wanted to resurrect the object, it could, but we know it doesn't so everything is fine.

So, my worry is probably mainly philosophical. In practice there is no real issue with `WeakRef`. I just don't particularly like it as a primitive. It's easy to misuse, and encourages repeated polling instead of getting a callback from a registry, or getting automatically cleaned up by an `Expando`/`WeakMap`.

Reply:
@mraleph Your understanding is correct. The JS `WeakRef` only observably clears if its referent is dead over a turn of the event loop to limit sync observation of GC. (It's actually a little stronger: `deref`ing a WeakRef also ensures it's live during the current turn.)

My recollection was that there was also a lot of arguments around ease-of-reasoning as well in TC39. Synchronously observing WeakRefs being cleared means GC hazards are harder to distinguish. With the "only at turn boundaries" model, you know WeakRef state doesn't change across a synchronous call, so I guess the idea when reviewing code you only have to scan for `await` or Promises.

I don't have much of an opinion on whether Dart should carry this behavior over, but if it doesn't, Dart payloads compiled to the web might find themselves suddenly leaking if they poll via `deref`, since a non-null `deref` in a turn keeps it alive for the remainder of the turn.

I also see some discussion talking about resurrection. FYI JS `FinalizationRegistry` is post-mortem and async (the finalizers get queued up as m**a**crotasks.)

Reply:
Thanks @syg! 

> I also see some discussion talking about resurrection. FYI JS FinalizationRegistry is post-mortem and async (the finalizers get queued up as macrotasks.)

The resurrection is just mentioned as something we don't support and we don't want to support :-D

Reply:
> The resurrection is just mentioned as something we don't support and we don't want to support :-D

Great decision! Very cursed stuff. :)

Reply:
@mraleph WeakRef is important to my use case. With WeakRef + FinalizationRegistry I know how to create a WeakMap with iterable keys and values which is a convenient abstraction for memory leak detection tooling that tracks objects that are disposed but not GCed within N seconds.